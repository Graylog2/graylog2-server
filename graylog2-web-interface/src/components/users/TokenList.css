:local(.tokenNewNameLabel) {
    margin-top: 8px;
}
