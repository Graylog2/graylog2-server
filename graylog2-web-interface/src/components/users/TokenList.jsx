import PropTypes from 'prop-types';
import React from 'react';
import Immutable from 'immutable';

import ClipboardButton from 'components/common/ClipboardButton';
import { Row, Col, FormControl, ControlLabel, Button, Checkbox, ButtonGroup } from 'react-bootstrap';
import TableList from 'components/common/TableList';
import Spinner from 'components/common/Spinner';
import TokenListStyle from './TokenList.css';

class TokenList extends React.Component {
  static propTypes = {
    tokens: PropTypes.arrayOf(PropTypes.object),
    onDelete: PropTypes.func,
    onCreate: PropTypes.func,
    creatingToken: PropTypes.bool,
    deletingToken: PropTypes.string,
  };

  static defaultProps = {
    tokens: [],
    onDelete: () => {},
    onCreate: () => {},
    creatingToken: false,
    deletingToken: undefined,
  };

  constructor(props) {
    super(props);

    this.state = {
      token_name: '',
      hide_tokens: true,
    };

    this._onNewTokeChanged = this._onNewTokeChanged.bind(this);
    this._onShowTokensChanged = this._onShowTokensChanged.bind(this);
    this._createToken = this._createToken.bind(this);
    this.itemActionsFactory = this.itemActionsFactory.bind(this);
  }

  _onNewTokeChanged(event) {
    this.setState({ token_name: event.target.value });
  }

  _onShowTokensChanged(event) {
    this.setState({ hide_tokens: event.target.checked });
  }

  _deleteToken(token) {
    return () => {
      this.props.onDelete(token.token, token.name);
    };
  }

  _createToken(e) {
    this.props.onCreate(this.state.token_name);
    this.setState({ token_name: '' });
    e.preventDefault();
  }

  itemActionsFactory(token) {
    const deleteButton = this.props.deletingToken === token.token ? <Spinner text="Deleting..." /> : 'Delete';
    return (
      <ButtonGroup>
        <ClipboardButton title={'Copy to clipboard'} text={token.token} bsSize="xsmall" />
        <Button bsSize="xsmall"
                disabled={this.props.deletingToken === token.token}
                bsStyle="primary"
                onClick={this._deleteToken(token)}>
          {deleteButton}
        </Button>
      </ButtonGroup>
    );
  }

  render() {
    const submitButton = (this.props.creatingToken ? <Spinner text="Creating..." /> : 'Create Token');

    const createTokenForm = (
      <form onSubmit={this._createToken}>
        <div className="form-group">
          <Row>
            <Col sm={2}>
              <ControlLabel className={TokenListStyle.tokenNewNameLabel}>Token Name</ControlLabel>
            </Col>
            <Col sm={4}>
              <FormControl
                id="create-token-input"
                type="text"
                placeholder="e.g ServiceName"
                value={this.state.token_name}
                onChange={this._onNewTokeChanged} />
            </Col>
            <Col sm={2}>
              <Button id="create-token"
              disabled={this.state.token_name === '' || this.props.creatingToken}
              type="submit"
              bsStyle="primary" >{submitButton}</Button>
            </Col>
          </Row>
        </div>
        <hr />
      </form>);

    return (
      <span>
        {createTokenForm}
        <TableList filterKeys={['name', 'token']}
          items={Immutable.List(this.props.tokens)}
          idKey="token"
          titleKey="name"
          descriptionKey="token"
          hideDescription={this.state.hide_tokens}
          enableBulkActions={false}
          itemActionsFactory={this.itemActionsFactory} />
        <Checkbox id="hide-tokens" onChange={this._onShowTokensChanged} checked={this.state.hide_tokens}>
          Hide Tokens
        </Checkbox>
      </span>
    );
  }
}

export default TokenList;
