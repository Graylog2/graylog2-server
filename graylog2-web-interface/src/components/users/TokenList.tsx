/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import PropTypes from 'prop-types';
import React, { useMemo, useState } from 'react';
import styled from 'styled-components';

import { ClipboardButton, ControlledTableList, Icon, Timestamp, SearchForm, Spinner } from 'components/common';
import { Button, Col, Panel, Row } from 'components/graylog';
import type { Token, TokenSummary } from 'actions/users/UsersActions';
import { sortByDate } from 'util/SortUtils';

import CreateTokenForm from './CreateTokenForm';

const StyledTokenPanel = styled(Panel)`
  &.panel {
    margin: 10px 0;
    background-color: ${(props) => props.theme.colors.global.contentBackground};

    .panel-heading {
      color: ${(props) => props.theme.colors.gray[30]};
    }
  }
`;

const StyledCopyTokenButton = styled(ClipboardButton)`
  vertical-align: baseline;
  margin-left: 1em;
`;

const StyledSearchForm = styled(SearchForm)`
  margin-bottom: 10px;
`;

const StyledLastAccess = styled.div`
  color: ${(props) => props.theme.colors.gray[60]};
  font-size: ${(props) => props.theme.fonts.size.small};
  margin-bottom: 5px;
`;

type Props = {
  creatingToken: boolean,
  deletingToken?: string,
  onCreate: (tokenName: string) => Promise<Token>,
  onDelete: (tokenId: string, tokenName: string) => void,
  tokens: TokenSummary[],
};

const TokenList = ({ creatingToken, deletingToken, onCreate, onDelete, tokens }: Props) => {
  const [createdToken, setCreatedToken] = useState<Token | undefined>();
  const [query, setQuery] = useState('');

  const effectiveTokens = useMemo(() => {
    const queryRegex = new RegExp(query, 'i');

    return tokens
      .filter(({ name }) => queryRegex.test(name))
      .sort((token1, token2) => sortByDate(token1.last_access, token2.last_access, 'desc'));
  }, [query, tokens]);

  const handleTokenCreation = (tokenName) => {
    const promise = onCreate(tokenName);

    promise.then((token) => {
      setCreatedToken(token);

      return token;
    });
  };

  const deleteToken = (token) => {
    return () => {
      onDelete(token.id, token.name);
    };
  };

  const updateQuery = (nextQuery?: string) => setQuery(nextQuery || '');

  return (
    <span>
      <CreateTokenForm onCreate={handleTokenCreation} creatingToken={creatingToken} />
      {createdToken && (
        <StyledTokenPanel bsStyle="success">
          <Panel.Heading>
            <Panel.Title>Token <em>{createdToken.name}</em> created!</Panel.Title>
          </Panel.Heading>
          <Panel.Body>
            <p>This is your new token. Make sure to copy it now, you will not be able to see it again.</p>
            <pre>
              {createdToken.token}
              <StyledCopyTokenButton title={<Icon name="clipboard" fixedWidth />} text={createdToken.token} bsSize="xsmall" />
            </pre>
            <Button bsStyle="primary" onClick={() => setCreatedToken(undefined)}>Done</Button>
          </Panel.Body>
        </StyledTokenPanel>
      )}
      <hr />
      <StyledSearchForm onSearch={updateQuery}
                        onReset={updateQuery}
                        searchButtonLabel="Find"
                        searchBsStyle="info"
                        label="Filter"
                        useLoadingState={false} />

      <ControlledTableList>
        <ControlledTableList.Header />
        {effectiveTokens.length === 0 && (
          <ControlledTableList.Item>
            <p>{query === '' ? 'No tokens to display.' : 'No tokens match the filter.'}</p>
          </ControlledTableList.Item>
        )}
        {effectiveTokens.map((token) => {
          const tokenNeverUsed = Date.parse(token.last_access) === 0;

          return (
            <ControlledTableList.Item key={token.id}>
              <Row className="row-sm">
                <Col md={9}>
                  {token.name}
                  <StyledLastAccess>
                    {tokenNeverUsed ? 'Never used' : <>Last used <Timestamp dateTime={token.last_access} relative /></>}
                  </StyledLastAccess>
                </Col>
                <Col md={3} className="text-right">
                  <Button bsSize="xsmall"
                          disabled={deletingToken === token.id}
                          bsStyle="primary"
                          onClick={deleteToken(token)}>
                    {deletingToken === token.id ? <Spinner text="Deleting..." /> : 'Delete'}
                  </Button>
                </Col>
              </Row>
            </ControlledTableList.Item>
          );
        })}
      </ControlledTableList>
    </span>
  );
};

TokenList.propTypes = {
  tokens: PropTypes.arrayOf(PropTypes.object),
  onDelete: PropTypes.func,
  onCreate: PropTypes.func,
  creatingToken: PropTypes.bool,
  deletingToken: PropTypes.string,
};

TokenList.defaultProps = {
  tokens: [],
  onDelete: () => {},
  onCreate: () => {},
  creatingToken: false,
  deletingToken: undefined,
};

export default TokenList;
