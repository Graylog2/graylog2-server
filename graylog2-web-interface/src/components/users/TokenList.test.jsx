import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import 'helpers/mocking/react-dom_mock';

import TokenList from 'components/users/TokenList';

jest.mock('components/common/ClipboardButton', () => 'clipboard-button');

describe('<TokenList />', () => {
  const tokens = [
    { name: 'Acme', token: 'beef2001' },
    { name: 'Hamfred', token: 'beef2002' },
  ];

  it('should render with empty tokens', () => {
    const wrapper = renderer.create(<TokenList tokens={[]} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with tokens', () => {
    const wrapper = renderer.create(<TokenList tokens={tokens} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should add new token and delete existing ones', () => {
    const createFn = jest.fn((tokenName) => {
      expect(tokenName).toEqual('hans');
    });
    const deleteFn = jest.fn((token) => {
      expect(token).toEqual('beef2001');
    });
    const wrapper = mount(<TokenList
      tokens={tokens}
      onCreate={createFn}
      onDelete={deleteFn}
    />);
    wrapper.find('input#create-token-input').simulate('change', { target: { value: 'hans' } });
    wrapper.find('form').at(0).simulate('submit');
    expect(createFn.mock.calls.length).toBe(1);

    wrapper.find('button[children="Delete"]').at(0).simulate('click');
    expect(createFn.mock.calls.length).toBe(1);
  });

  it('should display tokens if "Hide tokens" was unchecked', () => {
    const wrapper = mount(<TokenList
      tokens={tokens}
    />);
    expect(wrapper.find('span[children="beef2001"]').length).toEqual(0);
    wrapper.find('input#hide-tokens').simulate('change', { target: { checked: false } });
    expect(wrapper.find('span[children="beef2001"]').length).toEqual(1);
  });
});
