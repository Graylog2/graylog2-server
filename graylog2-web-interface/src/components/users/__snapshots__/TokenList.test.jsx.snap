// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenList /> should render with empty tokens 1`] = `
<span>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-2"
        >
          <label
            className="tokenNewNameLabel control-label"
          >
            Token Name
          </label>
        </div>
        <div
          className="col-sm-4"
        >
          <input
            className="form-control"
            id="create-token-input"
            onChange={[Function]}
            placeholder="e.g ServiceName"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-sm-2"
        >
          <button
            className="btn btn-primary"
            disabled={true}
            id="create-token"
            type="submit"
          >
            Create Token
          </button>
        </div>
      </div>
    </div>
    <hr />
  </form>
  <div>
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <div
          className="filter"
        >
          <form
            className="form-inline"
            onSubmit={[Function]}
            style={
              Object {
                "display": "inline",
              }
            }
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="filter-data-filter"
              >
                Filter
              </label>
              <div
                className="typeahead-wrapper"
              >
                <input
                  className="form-control"
                  id="filter-data-filter"
                  label="Filter"
                  placeholder=""
                  type="text"
                />
                
              </div>
            </div>
            <button
              className="btn btn-default"
              disabled={false}
              style={
                Object {
                  "marginLeft": 5,
                }
              }
              type="submit"
            >
              Filter
            </button>
            <button
              className="btn btn-default"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "marginLeft": 5,
                }
              }
              type="button"
            >
              Reset
            </button>
          </form>
          <ul
            className="pill-list"
          />
        </div>
      </div>
    </div>
    <div>
      <div
        className="list-group"
      >
        <span
          className="listGroupHeader list-group-item"
        >
          <div
            className="headerWrapper"
          >
            
          </div>
        </span>
        <span
          className="list-group-item"
        >
          No items to display
        </span>
      </div>
    </div>
  </div>
  <div
    className="checkbox"
  >
    <label
      title=""
    >
      <input
        checked={true}
        disabled={false}
        id="hide-tokens"
        onChange={[Function]}
        type="checkbox"
      />
      Hide Tokens
    </label>
  </div>
</span>
`;

exports[`<TokenList /> should render with tokens 1`] = `
<span>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-2"
        >
          <label
            className="tokenNewNameLabel control-label"
          >
            Token Name
          </label>
        </div>
        <div
          className="col-sm-4"
        >
          <input
            className="form-control"
            id="create-token-input"
            onChange={[Function]}
            placeholder="e.g ServiceName"
            type="text"
            value=""
          />
        </div>
        <div
          className="col-sm-2"
        >
          <button
            className="btn btn-primary"
            disabled={true}
            id="create-token"
            type="submit"
          >
            Create Token
          </button>
        </div>
      </div>
    </div>
    <hr />
  </form>
  <div>
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <div
          className="filter"
        >
          <form
            className="form-inline"
            onSubmit={[Function]}
            style={
              Object {
                "display": "inline",
              }
            }
          >
            <div
              className="form-group"
            >
              <label
                className="control-label"
                htmlFor="filter-data-filter"
              >
                Filter
              </label>
              <div
                className="typeahead-wrapper"
              >
                <input
                  className="form-control"
                  id="filter-data-filter"
                  label="Filter"
                  placeholder=""
                  type="text"
                />
                
              </div>
            </div>
            <button
              className="btn btn-default"
              disabled={false}
              style={
                Object {
                  "marginLeft": 5,
                }
              }
              type="submit"
            >
              Filter
            </button>
            <button
              className="btn btn-default"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "marginLeft": 5,
                }
              }
              type="button"
            >
              Reset
            </button>
          </form>
          <ul
            className="pill-list"
          />
        </div>
      </div>
    </div>
    <div>
      <div
        className="list-group"
      >
        <span
          className="listGroupHeader list-group-item"
        >
          <div
            className="headerWrapper"
          >
            
          </div>
        </span>
        <span
          className="list-group-item"
        >
          <div
            className="itemWrapper itemWrapperStatic"
          >
            <div
              className="itemActionsWrapper"
            >
              <div
                className="btn-group"
              >
                <clipboard-button
                  bsSize="xsmall"
                  text="beef2001"
                  title="Copy to clipboard"
                />
                <button
                  className="btn btn-xs btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
            <div
              className="header"
              id="token-input"
            >
              Acme
            </div>
          </div>
        </span>
        <span
          className="list-group-item"
        >
          <div
            className="itemWrapper itemWrapperStatic"
          >
            <div
              className="itemActionsWrapper"
            >
              <div
                className="btn-group"
              >
                <clipboard-button
                  bsSize="xsmall"
                  text="beef2002"
                  title="Copy to clipboard"
                />
                <button
                  className="btn btn-xs btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
            <div
              className="header"
              id="token-input"
            >
              Hamfred
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
  <div
    className="checkbox"
  >
    <label
      title=""
    >
      <input
        checked={true}
        disabled={false}
        id="hide-tokens"
        onChange={[Function]}
        type="checkbox"
      />
      Hide Tokens
    </label>
  </div>
</span>
`;
